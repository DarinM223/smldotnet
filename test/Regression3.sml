structure Regression3 = struct
  fun main () =
    let
      val f = ref ()
      val g = ref (print "hello\n")
      val () = f := !g
      val g = ref ((), 1, ())
      val () = g := ((), 2, ())
      val h = ref { a = (), b = 1, c = () }
      val () = h := { a = (), b = 2, c = (print "world\n"; ()) }
    in
      ()
    end
end