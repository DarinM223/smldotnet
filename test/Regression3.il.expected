.assembly extern mscorlib {
  .publickeytoken = (B7 7A 5C 56 19 34 E0 89 )
  .ver 4:0:0:0
}
.assembly extern Accessibility {
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )
  .ver 2:0:0:0
}
.assembly extern System.Configuration.Install {
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )
  .ver 2:0:0:0
}
.assembly extern System.Data {
  .publickeytoken = (B7 7A 5C 56 19 34 E0 89 )
  .ver 2:0:0:0
}
.assembly extern System.DirectoryServices {
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )
  .ver 2:0:0:0
}
.assembly extern System {
  .publickeytoken = (B7 7A 5C 56 19 34 E0 89 )
  .ver 2:0:0:0
}
.assembly extern System.Drawing {
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )
  .ver 2:0:0:0
}
.assembly extern System.EnterpriseServices {
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )
  .ver 2:0:0:0
}
.assembly extern System.Management {
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )
  .ver 2:0:0:0
}
.assembly extern System.Messaging {
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )
  .ver 2:0:0:0
}
.assembly extern System.Runtime.Remoting {
  .publickeytoken = (B7 7A 5C 56 19 34 E0 89 )
  .ver 2:0:0:0
}
.assembly extern System.Runtime.Serialization.Formatters.Soap {
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )
  .ver 2:0:0:0
}
.assembly extern System.Security {
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )
  .ver 2:0:0:0
}
.assembly extern System.ServiceProcess {
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )
  .ver 2:0:0:0
}
.assembly extern Mono.Unix {
  .publickeytoken = (CC 7B 13 FF CD 2D DD 51 )
  .ver 7:1:0:0
}
.assembly 'Regression3'
{ .hash algorithm 0x00008004 }
.module 'Regression3.exe'

.namespace '$' {
.class  private sealed 'Tuple1_a' extends ['mscorlib']'System'.'Object'{

.field assembly int32 '_1'
.method assembly instance void '.ctor'(int32 '_1') cil {
.maxstack 2
.zeroinit
  ldarg.0
  call instance void ['mscorlib']'System'.'Object'::'.ctor'()
  ldarg.0
  ldarg.1
  stfld int32 '$'.'Tuple1_a'::'_1'
  ret
{.locals ()}
}

}
}

.namespace '$' {
.class  private sealed 'Tuple1_b' extends ['mscorlib']'System'.'Object'{

.field assembly class '$'.'Tuple1_a' '_1'
.method assembly instance void '.ctor'(class '$'.'Tuple1_a' '_1') cil {
.maxstack 2
.zeroinit
  ldarg.0
  call instance void ['mscorlib']'System'.'Object'::'.ctor'()
  ldarg.0
  ldarg.1
  stfld class '$'.'Tuple1_a' '$'.'Tuple1_b'::'_1'
  ret
{.locals ()}
}

}
}

.namespace '$' {
.class  private abstract 'Class_a' extends ['mscorlib']'System'.'Exception'{

.field private initonly string 's'
.method public instance void '.ctor'(string 's') cil {
.maxstack 2
.zeroinit
  ldarg.0
  ldarg.1
  call instance instance void ['mscorlib']'System'.'Exception'::'.ctor'(string)
  ldarg.0
  ldarg.1
  stfld string '$'.'Class_a'::'s'
  ret
{.locals ()}
}
.method public virtual instance string 'ToString'() cil {
.maxstack 3
.zeroinit
  ldarg.0
  callvirt instance string '$'.'Class_a'::'ExnMessage'()
  stloc.0
  ldsfld string '$'.'Globals'::'$l'
  stloc.1
  ldloc.1
  ldnull
  ceq
  stloc.2
  ldloc.2
  brtrue L_1
  ldloc.0
  ldloc.1
  call string ['mscorlib']'System'.'String'::'Concat'(string,string)
  stloc.0
  ldloc.0
  ret
L_1:
  ldloc.0
  ret
{.locals ([0] string /*ctcj 'PrimUtils_.prefix'*/ ,
         [1] string /*ctch*/ ,
         [2] bool /*ctez*/ )}
}
.method public virtual instance string 'ExnMessage'() cil {
.maxstack 2
.zeroinit
  ldarg.0
  callvirt instance string '$'.'Class_a'::'ExnName'()
  stloc.0
  ldloc.0
  ldstr ": "
  call string ['mscorlib']'System'.'String'::'Concat'(string,string)
  stloc.0
  ldarg.0
  callvirt instance string ['mscorlib']'System'.'Exception'::'get_Message'()
  stloc.1
  ldloc.0
  ldloc.1
  call string ['mscorlib']'System'.'String'::'Concat'(string,string)
  stloc.0
  ldloc.0
  ret
{.locals ([0] string /*ctcp ctcn ctcm*/ ,
         [1] string /*ctco*/ )}
}
.method public virtual instance string 'ExnName'() cil {
.maxstack 1
.zeroinit
  ldstr "SML exception"
  ret
{.locals ()}
}

}
}

.class  public sealed 'Regression3' extends ['mscorlib']'System'.'Object'{

.method public static void '.cctor'() cil {
.maxstack 0
.zeroinit
  call void '$'.'Globals'::'$'()
  ret
{.locals ()}
}
.method public static void 'main'() cil {
.maxstack 2
.zeroinit
.entrypoint
  ldstr bytearray (68 00 65 00 6C 00 6C 00 6F 00 0A 00 )
  call void '$'.'Globals'::'TextIO.print'(string)
  ldsfld class '$'.'Tuple1_a' '$'.'Globals'::'$ctep'
  newobj instance void '$'.'Tuple1_b'::.ctor(class '$'.'Tuple1_a')
  stloc.0
  ldloc.0
  ldsfld class '$'.'Tuple1_a' '$'.'Globals'::'$cteq'
  stfld class '$'.'Tuple1_a' '$'.'Tuple1_b'::'_1'
  ldsfld class '$'.'Tuple1_a' '$'.'Globals'::'$ctep'
  newobj instance void '$'.'Tuple1_b'::.ctor(class '$'.'Tuple1_a')
  stloc.0
  ldstr bytearray (77 00 6F 00 72 00 6C 00 64 00 0A 00 )
  call void '$'.'Globals'::'TextIO.print'(string)
  ldloc.0
  ldsfld class '$'.'Tuple1_a' '$'.'Globals'::'$cteq'
  stfld class '$'.'Tuple1_a' '$'.'Tuple1_b'::'_1'
  ret
{.locals ([0] class '$'.'Tuple1_b' /*'Regression3.h' 'Regression3.g'*/ )}
}

}

.namespace '$' {
.class  private 'Fun' extends ['mscorlib']'System'.'Object'{

.method assembly instance void '.ctor'() cil {
.maxstack 2
.zeroinit
  ldarg.0
  call instance void ['mscorlib']'System'.'Object'::'.ctor'()
  ret
{.locals ()}
}

}
}

.namespace '$' {
.class  private sealed 'Tuple3_a' extends ['mscorlib']'System'.'Object'{

.field assembly string '_1'
.field assembly class ['mscorlib']'System'.'Exception' '_2'
.field assembly string '_3'
.method assembly instance void '.ctor'(string '_1',class ['mscorlib']'System'.'Exception' '_2',string '_3') cil {
.maxstack 4
.zeroinit
  ldarg.0
  call instance void ['mscorlib']'System'.'Object'::'.ctor'()
  ldarg.0
  ldarg.1
  stfld string '$'.'Tuple3_a'::'_1'
  ldarg.0
  ldarg.2
  stfld class ['mscorlib']'System'.'Exception' '$'.'Tuple3_a'::'_2'
  ldarg.0
  ldarg.3
  stfld string '$'.'Tuple3_a'::'_3'
  ret
{.locals ()}
}

}
}

.namespace '!'.'IO' {
.class  private sealed 'Io' extends '$'.'Class_a'{

.field assembly class '$'.'Tuple3_a' '_1'
.method assembly instance void '.ctor'(class '$'.'Tuple3_a' '') cil {
.maxstack 3
.zeroinit
  ldarg.0
  ldnull
  call instance void '$'.'Class_a'::'.ctor'(string)
  ldarg.0
  ldarg.1
  stfld class '$'.'Tuple3_a' '!'.'IO'.'Io'::'_1'
  ret
{.locals ()}
}
.method public virtual instance string 'ExnName'() cil {
.maxstack 3
.zeroinit
  ldstr "IO.Io"
  ret
{.locals ()}
}

}
}

.namespace '!'.'General' {
.class  private sealed 'Fail' extends '$'.'Class_a'{

.method assembly instance void '.ctor'(string '') cil {
.maxstack 3
.zeroinit
  ldarg.0
  ldarg.1
  call instance void '$'.'Class_a'::'.ctor'(string)
  ret
{.locals ()}
}
.method public virtual instance string 'ExnName'() cil {
.maxstack 3
.zeroinit
  ldstr "General.Fail"
  ret
{.locals ()}
}

}
}

.namespace '$' {
.class  private 'Globals' extends ['mscorlib']'System'.'Object'{

.field static assembly string '$l'
.field static assembly int32 '$g'
.field static assembly string '$ctdb'
.field static assembly int32 '$ctdh'
.field static assembly class ['mscorlib']'System'.'IO'.'TextWriter' '$ctdk'
.field static assembly class '$'.'Tuple1_a' '$ctep'
.field static assembly class '$'.'Tuple1_a' '$cteq'
.method static assembly void '.cctor'() cil {
.maxstack 3
.zeroinit
  ldc.i4.s 11
  conv.u4
  stloc.0
  ldc.i4.1
  ldloc.0
  shl
  stloc.1
  ldloc.1
  ldc.i4.1
  sub.ovf
  pop
  ldc.i4.s 11
  ldc.i4.1
  sub.ovf
  stloc.1
  ldloc.1
  conv.u4
  stloc.0
  ldc.i4.1
  ldloc.0
  shl
  stloc.1
  ldloc.1
  ldc.i4.1
  sub.ovf
  stloc.1
  ldloc.1
  ldc.i4.s 52
  add.ovf
  stloc.2
  ldloc.2
  ldc.i4.1
  sub.ovf
  pop
  ldc.i8 0x0000000000000040
  conv.ovf.i4
  stloc.2
  ldloc.2
  ldc.i4.1
  sub.ovf
  stloc.2
  ldloc.2
  ldloc.1
  add.ovf
  pop
  call string ['mscorlib']'System'.'Environment'::'get_NewLine'()
  stsfld string '$'.'Globals'::'$ctdb'
  ldsfld string '$'.'Globals'::'$ctdb'
  ldnull
  ceq
  stloc.3
  ldloc.3
  brtrue L_48
  ldsfld string '$'.'Globals'::'$ctdb'
  ldc.i4.s -2147483648
  callvirt instance char ['mscorlib']'System'.'String'::'get_Chars'(int32)
  stloc.s 4
  ldsfld string '$'.'Globals'::'$ctdb'
  callvirt instance int32 ['mscorlib']'System'.'String'::'get_Length'()
  stloc.1
  ldloc.1
  dup
  ldc.i4.1
  bne.un L_46
  pop
  ldloc.s 4
  ldc.i4 10
  ceq
  stloc.3
  ldloc.3
  brtrue L_45
  ldloc.s 4
  ldc.i4.s -1
  ldc.i4.1
  stsfld int32 '$'.'Globals'::'$ctdh'
  pop
  pop
  br _ctdg
L_45:
  ldc.i4.s -1
  ldc.i4.s -1
  ldc.i4.s -2147483648
  stsfld int32 '$'.'Globals'::'$ctdh'
  pop
  pop
  br _ctdg
L_46:
  ldc.i4.2
  bne.un L_47
  ldsfld string '$'.'Globals'::'$ctdb'
  ldc.i4.1
  callvirt instance char ['mscorlib']'System'.'String'::'get_Chars'(int32)
  stloc.s 5
  ldloc.s 4
  ldloc.s 5
  ldc.i4.2
  stsfld int32 '$'.'Globals'::'$ctdh'
  pop
  pop
  br _ctdg
L_47:
  ldstr " at TextIO:125.26-79"
  stsfld string '$'.'Globals'::'$l'
  ldstr "Unexpected line separator length"
  newobj instance void '!'.'General'.'Fail'::.ctor(string)
  throw
_ctdg:
  call class ['mscorlib']'System'.'IO'.'TextReader' ['mscorlib']'System'.'Console'::'get_In'()
  pop
  call class ['mscorlib']'System'.'IO'.'TextWriter' ['mscorlib']'System'.'Console'::'get_Out'()
  stsfld class ['mscorlib']'System'.'IO'.'TextWriter' '$'.'Globals'::'$ctdk'
  call class ['mscorlib']'System'.'IO'.'TextWriter' ['mscorlib']'System'.'Console'::'get_Error'()
  pop
  ldc.i4.1
  newobj instance void '$'.'Tuple1_a'::.ctor(int32)
  stsfld class '$'.'Tuple1_a' '$'.'Globals'::'$ctep'
  ldc.i4.2
  newobj instance void '$'.'Tuple1_a'::.ctor(int32)
  stsfld class '$'.'Tuple1_a' '$'.'Globals'::'$cteq'
  ret
L_48:
  ldstr " at TextIO:118.20-68"
  stsfld string '$'.'Globals'::'$l'
  ldstr "No line separator available"
  newobj instance void '!'.'General'.'Fail'::.ctor(string)
  throw
{.locals ([0] unsigned int32 /*ctcv ctcs*/ ,
         [1] int32 /*ctdf 'Real.eoffset' ctcw ctcu ctct*/ ,
         [2] int32 /*ctda 'Real.large_precision' 'Real.meoffset'*/ ,
         [3] bool /*ctev ctey*/ ,
         [4] char /*'TextIO.c'*/ ,
         [5] char /*ctew*/ )}
}
.method static assembly void '$'() cil {
.maxstack 0
.zeroinit
  ret
{.locals ()}
}
.method static assembly void '.GCSafePoint'() managed noinlining {
.maxstack 0
.zeroinit
  ret
{.locals ()}
}
.method static assembly void 'TextIO.print'(string 'ctdm') cil {
.maxstack 3
.zeroinit
  ldsfld int32 '$'.'Globals'::'$ctdh'
  ldc.i4.s -2147483648
  ceq
  stloc.0
  ldloc.0
  brtrue L_34
  ldarg.0
  callvirt instance int32 ['mscorlib']'System'.'String'::'get_Length'()
  stloc.1
  ldc.i4.s -2147483648
  stloc.3
_ctdt:
_ctdu:
  ldloc.3
  ldloc.1
  clt
  stloc.0
  ldloc.0
  brtrue L_7
L_2:
  ldsfld class ['mscorlib']'System'.'IO'.'TextWriter' '$'.'Globals'::'$ctdk'
  callvirt instance void ['mscorlib']'System'.'IO'.'TextWriter'::'Flush'()
  leave L_3
L_3:
  ret
L_5:
  stloc.2
  ldstr " at TextIO:87.8-91.12"
  stsfld string '$'.'Globals'::'$l'
  ldstr "TextWriter"
  ldloc.2
  ldstr "flushOut"
  newobj instance void '$'.'Tuple3_a'::.ctor(string,class ['mscorlib']'System'.'Exception',string)
  newobj instance void '!'.'IO'.'Io'::.ctor(class '$'.'Tuple3_a')
  throw
L_6:
L_4:
  ret
.try L_2 to L_3 catch ['mscorlib']'System'.'Exception' handler L_5 to L_6
L_7:
L_8:
  ldarg.0
  ldloc.3
  callvirt instance char ['mscorlib']'System'.'String'::'get_Chars'(int32)
  stloc.s 4
  leave L_9
L_9:
  ldloc.s 4
  ldc.i4 10
  ceq
  stloc.0
  ldloc.0
  brtrue L_20
L_10:
  ldsfld class ['mscorlib']'System'.'IO'.'TextWriter' '$'.'Globals'::'$ctdk'
  ldloc.s 4
  callvirt instance void ['mscorlib']'System'.'IO'.'TextWriter'::'Write'(char)
  leave L_11
L_11:
L_12:
  ldloc.3
  ldc.i4.1
  add.ovf
  stloc.3
  leave L_13
L_13:
  br _ctdu
L_15:
  stloc.2
  leave _ctdp
L_16:
L_14:
  ret
.try L_12 to L_13 catch ['mscorlib']'System'.'Exception' handler L_15 to L_16
L_18:
  stloc.2
  leave _ctdp
L_19:
L_17:
  ret
.try L_10 to L_11 catch ['mscorlib']'System'.'Exception' handler L_18 to L_19
L_20:
L_21:
  ldsfld class ['mscorlib']'System'.'IO'.'TextWriter' '$'.'Globals'::'$ctdk'
  ldsfld string '$'.'Globals'::'$ctdb'
  callvirt instance void ['mscorlib']'System'.'IO'.'TextWriter'::'Write'(string)
  leave L_22
L_22:
L_23:
  ldloc.3
  ldc.i4.1
  add.ovf
  stloc.3
  leave L_24
L_24:
  br _ctdu
L_26:
  stloc.2
  leave _ctdp
L_27:
L_25:
  ret
.try L_23 to L_24 catch ['mscorlib']'System'.'Exception' handler L_26 to L_27
L_29:
  stloc.2
  leave _ctdp
L_30:
L_28:
  ret
.try L_21 to L_22 catch ['mscorlib']'System'.'Exception' handler L_29 to L_30
L_32:
  stloc.2
  leave _ctdp
L_33:
L_31:
  ret
.try L_8 to L_9 catch ['mscorlib']'System'.'Exception' handler L_32 to L_33
_ctdp:
  ldstr " at TextIO:87.8-91.12"
  stsfld string '$'.'Globals'::'$l'
  ldstr "TextWriter"
  ldloc.2
  ldstr "output"
  newobj instance void '$'.'Tuple3_a'::.ctor(string,class ['mscorlib']'System'.'Exception',string)
  newobj instance void '!'.'IO'.'Io'::.ctor(class '$'.'Tuple3_a')
  throw
L_34:
L_35:
  ldsfld class ['mscorlib']'System'.'IO'.'TextWriter' '$'.'Globals'::'$ctdk'
  ldarg.0
  callvirt instance void ['mscorlib']'System'.'IO'.'TextWriter'::'Write'(string)
  leave L_36
L_36:
L_37:
  ldsfld class ['mscorlib']'System'.'IO'.'TextWriter' '$'.'Globals'::'$ctdk'
  callvirt instance void ['mscorlib']'System'.'IO'.'TextWriter'::'Flush'()
  leave L_38
L_38:
  ret
L_40:
  stloc.2
  ldstr " at TextIO:87.8-91.12"
  stsfld string '$'.'Globals'::'$l'
  ldstr "TextWriter"
  ldloc.2
  ldstr "flushOut"
  newobj instance void '$'.'Tuple3_a'::.ctor(string,class ['mscorlib']'System'.'Exception',string)
  newobj instance void '!'.'IO'.'Io'::.ctor(class '$'.'Tuple3_a')
  throw
L_41:
L_39:
  ret
.try L_37 to L_38 catch ['mscorlib']'System'.'Exception' handler L_40 to L_41
L_43:
  stloc.2
  ldstr " at TextIO:87.8-91.12"
  stsfld string '$'.'Globals'::'$l'
  ldstr "TextWriter"
  ldloc.2
  ldstr "output"
  newobj instance void '$'.'Tuple3_a'::.ctor(string,class ['mscorlib']'System'.'Exception',string)
  newobj instance void '!'.'IO'.'Io'::.ctor(class '$'.'Tuple3_a')
  throw
L_44:
L_42:
  ret
.try L_35 to L_36 catch ['mscorlib']'System'.'Exception' handler L_43 to L_44
{.locals ([0] bool /*ctec ctdw ctdn*/ ,
         [1] int32 /*'CharVector.stop'*/ ,
         [2] class ['mscorlib']'System'.'Exception' /*ctej ctem ctea cteg cteh cted ctee ctdx ctdq*/ ,
         [3] int32 /*ctei ctef ctdv*/ ,
         [4] char /*cteb*/ )}
}

}
}
